### PR TITLE
Add symlink support

### LINKED ISSUE #194: VFS symlinks
The virtual filesystem should support files that mirror other files, similar to the way that directory mounts work.

### PR DESCRIPTION
**Pull request by SamTebbs33**
Closes #194﻿

### CODE REVIEW

File: src/kernel/filesystem/vfs.zig [hunk 1]
```diff
@@ -261,15 +298,23 @@ fn traversePath(path: []const u8, flags: OpenFlags) (Allocator.Error || Error)!*
             }
             return switch (node.*) {
                 .File => Error.NotADirectory,
-                .Dir => |*dir| try func(split, try dir.open(seg, rec_flags), rec_flags),
+                .Dir => |*dir| blk: {
+                    var n = try dir.open(seg, rec_flags, .{});
```
Comment (DrDeano) [inline src/kernel/filesystem/vfs.zig]: Maybe `n` isn't the best variable name lol

File: src/kernel/filesystem/vfs.zig [hunk 2]
```diff
@@ -370,19 +428,34 @@ pub fn openFile(path: []const u8, flags: OpenFlags) (Allocator.Error || Error)!*
...
             file.close();
             break :blk Error.NotADirectory;
         },
+        // We instructed open to folow symlinks above, so this is impossible
+        .Symlink => unreachable,
         .Dir => &node.Dir,
     };
 }
 
+pub fn openSymlink(path: []const u8, target: ?[]const u8, flags: OpenFlags) (Allocator.Error || Error)![]const u8 {
```
Comment (DrDeano) [inline src/kernel/filesystem/vfs.zig:465]: Doc comment

**Review by DrDeano — COMMENTED** [review]
For the initrd, you could add `.CREATE_SYMLINK` test for opening files and directories in the `open fail with invalid flags` test. Also, in the vfs.open implemention, the follow_sym is true, but in the initFS it is false, is there a reason?

**Comment by SamTebbs33** [comment]
> Also, in the vfs.open implemention, the follow_sym is true, but in the initFS it is false, is there a reason?

Do you mean that it is true in `openSymlink`? What is `initFS`?

**Comment by DrDeano** [comment]
Sorry, I mean `InitrdFS`, and in the tests for InitrdFS, where `open` is used, the `follow_sym` is set to `false` but in the `OpenFile` and `OpenDir` in `vfs.zig`, the `follow_sym` is set to `true`.

**Comment by SamTebbs33** [comment]
> Sorry, I mean `InitrdFS`, and in the tests for InitrdFS, where `open` is used, the `follow_sym` is set to `false` but in the `OpenFile` and `OpenDir` in `vfs.zig`, the `follow_sym` is set to `true`.

Oh right. I used `false` since I knew there wouldn't be any symlinks and so it didn't _have_ to be true, but the behaviour is the same if `true` is used, so I can change it to that. That could make it more reliable in the future too.

**Review by DrDeano — APPROVED** [review]

**Pull request merged**